src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/TypeClass.java

```diff
@@ -58,15 +61,38 @@ static boolean isRegisterAggregate(MemoryLayout type) {
         return type.bitSize() <= MAX_AGGREGATE_REGS_SIZE * 64;
     }
 
+    static List<MemoryLayout> scalarLayouts(GroupLayout gl) {
+        List<MemoryLayout> out = new ArrayList<>();
+        scalarLayoutsR(out, gl);
+        return out;
+    }
+
+    private static void scalarLayoutsR(List<MemoryLayout> out, GroupLayout gl) {
+        for (MemoryLayout member : gl.memberLayouts()) {
+            if (member instanceof GroupLayout memberGl) {
+                scalarLayoutsR(out, memberGl);
+            } else if (member instanceof SequenceLayout memberSl) {
+                for (long i = 0; i < memberSl.elementCount(); i++) {
+                    out.add(memberSl.elementLayout());
+                }
+            } else {
+                // padding or value layouts
+                out.add(member);
```

Comment:
Note here that groups with padding will be rejected later when each field is checked in `isHomogeneousFloatAggregate`

Reply:
Does the order in which you collect the layouts matter? E.g. this will do a depth-first visit of the layout tree - I see that the code above does an indexed get on the scalar layout list, so I believe the order does matter.

My impression is that it's ok: the depth-first visit returns layouts in the order in which they are laid out in memory, which seems to make sense for an ABI.

Reply:
It doesn't matter for the classification, all the ValueLayouts just need to be the same. Later when computing the struct storages in CallArranger, it also doesn't really matter. There we look at the carrier, byte size, and maybe alignment. But since all the layouts are again the same, they could technically be processed in any order.